test/subscription_map.hpp

```diff
@@ -438,32 +438,36 @@ class multiple_subscription_map
 
     // Insert a key => value at the specified subscription path
     // returns the handle and true if key was inserted, false if key was updated
-    template <typename V>
-    std::pair<handle, bool> insert_or_update(MQTT_NS::string_view subscription, Key const &key, V&& value) {
+    template <typename P>
+    std::pair<handle, bool> insert_or_update(MQTT_NS::string_view subscription, P&& new_pair) {
         auto path = this->find_subscription(subscription);
         if (path.empty()) {
             auto new_subscription_path = this->create_subscription(subscription);
-            new_subscription_path.back()->second.value[key] = std::forward<V>(value);
+            new_subscription_path.back()->second.value.insert(std::move(new_pair));
             ++this->map_size;
             return std::make_pair(this->path_to_handle(new_subscription_path), true);
         }
         else {
-            auto new_pair = std::make_pair(key, std::forward<V>(value));
-            auto insert_result = path.back()->second.value.insert(new_pair);
+            auto insert_result = path.back()->second.value.insert(std::move(new_pair));
             if(insert_result.second) {
                 this->increase_subscriptions(path);
                 ++this->map_size;
             } else {
-                insert_result.first->second = new_pair.second;
+                insert_result.first->second = std::move(new_pair.second);
```

Comment:
I couldn't find solution so far. But std::move with forwarding reference is dangerous.
It is expected behavior. 
See https://wandbox.org/permlink/fDcwUs7CLfNKTZLW
The second case, p is not modified.

But if you use std::move for the second member, it is modified unexpectedly.
https://wandbox.org/permlink/ASVYXBYO2vMRVr3T
See the second case.

Reply:
> I couldn't find solution so far. But std::move with forwarding reference is dangerous.
> It is expected behavior.
> See https://wandbox.org/permlink/fDcwUs7CLfNKTZLW
> The second case, p is not modified.
> 
> But if you use std::move for the second member, it is modified unexpectedly.
> https://wandbox.org/permlink/ASVYXBYO2vMRVr3T
> See the second case.

I see, i would have expected for unordered map to behave the same way, but it does not, why ?
https://wandbox.org/permlink/c6st3KcBX0fJPTrZ

Reply:
I come up with a solution. Replace forwarding reference with passed by value.
See https://wandbox.org/permlink/GG42y8PAJadSp5BL

One additional move constructor call is required, but it is usually very small cost.

Reply:
I tried similar way. I strongly recommend that start with simple case using wandbox or something.

```cpp
using rr_t = typename std::remove_reference<P>::type;
using sec_t = typename rr_t::second_type;
... = std::forward<sec_t>(p.second);
```

But it doesn't work expectedly.

Reply:
I'm not sure forwarding reference and `std::forward<T>(t)` work for a part (member) of forwarding reference. Perhaps asking stackoverflow is fast way to get answer. Today, I have exhausted. I'm going to sleep :)

Reply:
Good night, 

I think it is because the pair is a rvalue reference, but the individual members of the pair are normal references.

Reply:
https://wandbox.org/permlink/KtJDGMWmbqd9Lo9b

Reply:
C++ actually does not have an unordered_map insert_or_assign with a pair, I think for this reason:
https://en.cppreference.com/w/cpp/container/map/insert_or_assign